django/channels/database_layer.py

```diff
@@ -0,0 +1,245 @@
+        self._clean_expired()
+        # Get a message from one of our channels
+        while True:
+            try:
+                with transaction.atomic():
+                    message = self.channel_model.objects.filter(channel__in=channels).order_by("id").first()
+                    if message:
+                        self.channel_model.objects.filter(pk=message.pk).delete()
+                        return message.channel, self.deserialize(message.content)
+            except OperationalError:
+                # The database is probably trying to prevent a deadlock
+                time.sleep(0.1)
+                continue
+            if block:
+                time.sleep(1)
+            else:
+                return None, None
+
+    def new_channel(self, pattern):
+        assert isinstance(pattern, six.text_type)
+        assert pattern.endswith("!")
+        # Keep making channel names till one isn't present.
+        while True:
+            random_string = "".join(random.choice(string.ascii_letters) for i in range(10))
+            new_name = pattern + random_string
+            if not self.channel_model.objects.filter(channel=new_name).exists():
+                return new_name
+
+    # ASGI Group extension
+
+    def group_add(self, group, channel):
+        """
+        Adds the channel to the named group for at least 'expiry'
+        seconds (expiry defaults to message expiry if not provided).
+        """
+        self.group_model.objects.update_or_create(
+            group=group,
+            channel=channel,
+        )
+
+    def group_discard(self, group, channel):
+        """
+        Removes the channel from the named group if it is in the group;
+        does nothing otherwise (does not error)
+        """
+        self.group_model.objects.filter(group=group, channel=channel).delete()
+
+    def send_group(self, group, message):
+        """
+        Sends a message to the entire group.
+        """
+        self._clean_expired()
+        for channel in self.group_model.objects.filter(group=group).values_list("channel", flat=True):
+            self.send(channel, message)
+
+    # ASGI Flush extension
+
+    def flush(self):
+        self.channel_model.objects.all().delete()
+        self.group_model.objects.all().delete()
+
+    # Serialization
+
+    def serialize(self, message):
+        return AsgiJsonEncoder().encode(message)
+
+    def deserialize(self, message):
+        return AsgiJsonDecoder().decode(message)
+
+    # Database state mgmt
+
+    @property
+    def connection(self):
+        """
+        Returns the correct connection for the current thread.
+        """
+        return connections[self.db_alias]
+
+    @cached_property
+    def channel_model(self):
+        """
+        Initialises a new model to store messages; not done as part of a
+        models.py as we don't want to make it for most installs.
+        """
+        # Make the model class
+        class Message(models.Model):
```

Comment:
Honestly, maybe - initially I wanted to make this as easy to use as possible, but at this point the database backend is almost shipped as a fallback (I was even tempted to remove it entirely as it's so slow, but I figure we still have the `sqlite3` backend for databases for ease-of-use reasons)

Not sure how we could dynamically surface it in models.py based on settings, as you can technically instantiate the layers without having them done in a setting (e.g. during tests).